python_modules/dagster/dagster/_core/definitions/load_assets_from_modules.py

```diff
@@ -47,6 +49,90 @@ def find_subclasses_in_module(
+
+    @classmethod
+    def from_modules(cls, modules: Iterable[ModuleType]) -> "LoadedAssetsList":
+        return cls(
+            {
+                module.__name__: list(
+                    find_objects_in_module_of_types(
+                        module, (AssetsDefinition, SourceAsset, CacheableAssetsDefinition)
+                    )
+                )
+                for module in modules
+            },
+        )
+
+    @cached_property
+    def flat_object_list(self) -> Sequence[LoadableAssetTypes]:
+        return [
+            asset_object for objects in self.assets_per_module.values() for asset_object in objects
+        ]
+
+    @cached_property
+    def objects_by_id(self) -> Dict[int, LoadableAssetTypes]:
+        return {id(asset_object): asset_object for asset_object in self.flat_object_list}
+
+    @cached_property
+    def deduped_objects(self) -> Sequence[LoadableAssetTypes]:
+        return list(self.objects_by_id.values())
+
+    @cached_property
+    def module_name_by_id(self) -> Dict[int, str]:
+        return {
+            id(asset_object): module_name
+            for module_name, objects in self.assets_per_module.items()
+            for asset_object in objects
+        }
+
+    @cached_property
+    def objects_by_key(self) -> Mapping[AssetKey, Sequence[Union[SourceAsset, AssetsDefinition]]]:
+        objects_by_key = defaultdict(list)
+        for asset_object in self.flat_object_list:
+            if not isinstance(asset_object, KeyScopedAssetObjects):
+                continue
+            for key in key_iterator(asset_object):
+                objects_by_key[key].append(asset_object)
+        return objects_by_key
+
+    @cached_property
+    def sources(self) -> Sequence[SourceAsset]:
+        return [asset for asset in self.deduped_objects if isinstance(asset, SourceAsset)]
+
+    @cached_property
+    def assets_defs(self) -> Sequence[AssetsDefinition]:
+        return [asset for asset in self.deduped_objects if isinstance(asset, AssetsDefinition)]
+
+    @cached_property
+    def cacheables(self) -> Sequence[CacheableAssetsDefinition]:
+        return [
+            asset for asset in self.deduped_objects if isinstance(asset, CacheableAssetsDefinition)
+        ]
+
+    def _do_collision_detection(self) -> None:
+        for key, asset_objects in self.objects_by_key.items():
+            # If there is more than one asset_object in the list for a given key, and the objects do not refer to the same asset_object in memory, we have a collision.
+            num_distinct_objects_for_key = len(
+                set(id(asset_object) for asset_object in asset_objects)
+            )
+            if len(asset_objects) > 1 and num_distinct_objects_for_key > 1:
+                raise DagsterInvalidDefinitionError(
+                    f"Asset key {key.to_user_string()} is defined multiple times. Definitions found in modules: {', '.join(set(self.module_name_by_id[id(asset_object)]
```

Comment:
not critical (this is pre-existing thing), but one possibility here is for the asset to be defined twice in the same module, at which point this error message wouldn't be as helpful

wondering if in the case that the len(modules) == 1, we could print out the reprs of the individual objects or something.

feel free to ignore tho